### PR TITLE
[Exiled.API] Features.CustomStats

### PR DESCRIPTION
**Pull request by Misfiy**
# Additions
* ``Player::MaxHumeShield``
* ``Player::MaxStamina``

### CODE REVIEW

**Review by NaoUnderscore — COMMENTED** [review]
Does it require testing?

**Comment by Misfiy** [comment]
yyess probably
its basically the same as our thing for MaxHealth

**Comment by ALEXWARELLC** [comment]
Starting test now.

File: Exiled.API/Features/Player.cs [hunk 1]
```diff
@@ -907,9 +912,18 @@ public float HumeShield
         }
 
         /// <summary>
-        /// Gets the player's <see cref="PlayerStatsSystem.HumeShieldStat"/>.
+        /// Gets or sets the player's Max amount of Hume Shield.
```
Comment (Vladislav-CS) [inline Exiled.API/Features/Player.cs:915]:
```suggestion
        /// Gets or sets the player's maximum amount of Hume Shield.
```

File: Exiled.API/Features/CustomStats/CustomStaminaStat.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+// </copyright>
+// -----------------------------------------------------------------------
+
+namespace Exiled.API.Features.CustomStats
+{
+    using Mirror;
+    using PlayerStatsSystem;
+
+    /// <summary>
+    /// A custom version of <see cref="StaminaStat"/> which allows the player's max amount of Stamina to be changed.
```
Comment (Vladislav-CS) [inline Exiled.API/Features/CustomStats/CustomStaminaStat.cs:14]:
```suggestion
    /// A custom version of <see cref="StaminaStat"/> which allows the player's maximum amount of Stamina to be changed.
```

File: Exiled.API/Features/CustomStats/CustomStaminaStat.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    /// </summary>
+    public class CustomStaminaStat : StaminaStat
+    {
+        /// <inheritdoc/>
+        public override float MaxValue => CustomMaxValue == default ? base.MaxValue : CustomMaxValue;
+
+        /// <summary>
+        /// Gets or sets the maximum amount of health the player will have.
```
Comment (Vladislav-CS) [inline Exiled.API/Features/CustomStats/CustomStaminaStat.cs:22]:
```suggestion
        /// Gets or sets the maximum amount of stamina the player will have.
```

File: Exiled.API/Features/CustomStats/CustomStaminaStat.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            {
+                base.WriteValue(writer);
+                return;
+            }
+
+            writer.WriteByte(ToByte(CurValue / CustomMaxValue));
+        }
+
+        /// <summary>
+        /// Overiding NW Method to sync Player percentage of Stamina.
```
Comment (Vladislav-CS) [inline Exiled.API/Features/CustomStats/CustomStaminaStat.cs:42]:
```suggestion
        /// Overriding NW Method to sync Player percentage of Stamina.
```

File: Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    /// <summary>
+    /// A custom version of <see cref="HumeShieldStat"/> which allows the player's max amount of HumeShield to be changed.
+    /// </summary>
+    public class CustomHumeShieldStat : HumeShieldStat
+    {
+        /// <inheritdoc/>
+        public override float MaxValue => CustomMaxValue == default ? base.MaxValue : CustomMaxValue;
+
+        /// <summary>
+        /// Gets or sets the maximum amount of health the player will have.
```
Comment (Vladislav-CS) [inline Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs:25]:
```suggestion
        /// Gets or sets the maximum amount of HumeShield the player can have.
```

File: Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    using Mirror;
+    using PlayerRoles.PlayableScps.HumeShield;
+    using PlayerStatsSystem;
+    using UnityEngine;
+    using Utils.Networking;
+
+    /// <summary>
+    /// A custom version of <see cref="HumeShieldStat"/> which allows the player's max amount of HumeShield to be changed.
+    /// </summary>
+    public class CustomHumeShieldStat : HumeShieldStat
```
Comment (Vladislav-CS) [inline Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs:19]: Is it possible to change other properties of HumeShield?

**Review by Vladislav-CS — CHANGES_REQUESTED** [review]

File: Exiled.API/Features/Player.cs [hunk 7]
```diff
@@ -179,6 +182,8 @@ private set
                 CameraTransform = value.PlayerCameraReference;
 
                 value.playerStats._dictionarizedTypes[typeof(HealthStat)] = value.playerStats.StatModules[Array.IndexOf(PlayerStats.DefinedModules, typeof(HealthStat))] = healthStat = new CustomHealthStat { Hub = value };
+                value.playerStats._dictionarizedTypes[typeof(HumeShieldStat)] = value.playerStats.StatModules[Array.IndexOf(PlayerStats.DefinedModules, typeof(HumeShieldStat))] = humeShieldStat = new CustomHumeShieldStat { Hub = value };
+                value.playerStats._dictionarizedTypes[typeof(StaminaStat)] = value.playerStats.StatModules[Array.IndexOf(PlayerStats.DefinedModules, typeof(StaminaStat))] = staminaStat = new CustomStaminaStat { Hub = value };
```
Comment (ALEXWARELLC) [inline Exiled.API/Features/Player.cs:186]: Null Reference Execption is kinda cringe.

**Review by ALEXWARELLC — CHANGES_REQUESTED** [review]
Cringe.

File: Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs (hunk 6 above)
Comment (Misfiy) [reply Exiled.API/Features/CustomStats/CustomHumeShieldStat.cs:19]: image

File: Exiled.API/Features/Player.cs (hunk 7 above)
Comment (Misfiy) [reply Exiled.API/Features/Player.cs:186]: From what I saw in vc, this is an issue with Player.Get(ICommandSender) returning null

**Comment by ALEXWARELLC** [comment]
# Test Results
## Test Complete by @ALEXWARELLC 

No notes to show.

### Tester Notes
- While testing, SCP: SL decided to act stupidly, and started lagging back to 6 FPS. Take test with a grain of salt.

File: Exiled.API/Features/Player.cs (hunk 7 above)
Comment (ALEXWARELLC) [reply Exiled.API/Features/Player.cs:186]: Yep.

**Review by ALEXWARELLC — APPROVED** [review]
I'm ok with this. My requested changes was a me issue.

**Comment by ALEXWARELLC** [comment]
Will perform test later today.

**Comment by ALEXWARELLC** [comment]
Unable to test.

Reason: Yamato Skill Issue.

**Comment by ALEXWARELLC** [comment]
Cannot test as we cannot figure out what is causing a major lag issue.

**Comment by Misfiy** [comment]
This PR causes major lag. Not sure why.

**Pull request closed without merging**